lockfile_unix.go

```diff
@@ -24,38 +25,84 @@ func getLockFile(path string, ro bool) (Locker, error) {
 		return nil, errors.Wrapf(err, "error opening %q", path)
 	}
 	unix.CloseOnExec(fd)
+
+	locktype := unix.F_WRLCK
 	if ro {
-		return &lockfile{file: path, fd: uintptr(fd), lw: stringid.GenerateRandomID(), locktype: unix.F_RDLCK, locked: false}, nil
+		locktype = unix.F_RDLCK
 	}
-	return &lockfile{file: path, fd: uintptr(fd), lw: stringid.GenerateRandomID(), locktype: unix.F_WRLCK, locked: false}, nil
+	return &lockfile{
+		stateMutex: &sync.Mutex{},
+		writeMutex: &sync.Mutex{},
+		file:       path,
+		fd:         uintptr(fd),
+		lw:         stringid.GenerateRandomID(),
+		locktype:   int16(locktype),
+		locked:     false,
+		ro:         ro}, nil
 }
 
 type lockfile struct {
-	mu       sync.Mutex
-	file     string
-	fd       uintptr
-	lw       string
-	locktype int16
-	locked   bool
+	// stateMutex is used to synchronize concurrent accesses
+	stateMutex *sync.Mutex
+	// writeMutex is used to serialize and avoid recursive writer locks
+	writeMutex *sync.Mutex
+	counter    int64
+	file       string
+	fd         uintptr
+	lw         string
+	locktype   int16
+	locked     bool
+	ro         bool
 }
 
-// Lock locks the lock file
-func (l *lockfile) Lock() {
+// lock locks the lockfile via FCTNL(2) based on the specified type and
+// command.
+func (l *lockfile) lock(l_type int16, l_cmd int) {
```

Comment:
`l_cmd` is currently ignored.  Is it supposed to be?

Reply:
Wow, good catch! That's an artifact of an earlier version but not needed anymore as we're always setting the lock via `unix.F_SETLKW`. Will clean that up, thanks a lot!